### ISSUE TITLE
Query streamed with .stream() does not close when the stream has an error

### ISSUE THREAD

**Issue by mdierolf** [sidorares/node-mysql2#1168]
When streaming a query with pool.query.stream(), if .destroy() is called on the stream, the 'close' event does not propagate upstream and close the underlying connection, leaving dead connections in the pool which are still executing.

As such it seems best to forcibly close the connection as I do not think mysql has a way to cancel a query while leaving the connection open.

Please see quick fix below, I have not had time yet to write a test or to confirm that this is the proper way to close the connection, or whether it might have other side effects:
https://github.com/CloudQuote/node-mysql2/commit/ac9236c7638974c100c421a79ec516f5df3fa156

**Comment by sidorares** [comment]
if the problem is on the receiving end of the pipe not sure what is best, sometimes it's kill the connection but sometimes maybe just disconnecting pipe from it and letting it finish query ( and allow to process other commands )

**Comment by mdierolf** [comment]
The behavior can certainly be worked around, however if you use the stream.pipeline() function in node to transform the rows and pipe them downstream, the default cleanup behavior is to call .destroy() on every stream in the pipeline when any of them has an error.

For example:
`
import stream from 'stream';

stream.pipeline(
 pool.query("SELECT blah").stream(),
 myTransformStream,
 output
);
`

You could work around the destruction issue by creating the query outside the pipeline, and adding a .on() handler to the other two streams, but it defeats the purpose of the pipeline module, which is to simplify the piping and cleanup of streams.

Most users of node are rather hopeless when it comes to setting up a chain of streams and propagating every permutation of error properly, so I would think anything that allows the pipeline() function to be used would be a good idea.

Keep in mind, once the remaining streams in the pipeline are destroyed, their is nowhere to send the data to, and the connection will never be released back into the pool, eventually locking up the application

**Comment by sidorares** [comment]
> Keep in mind, once the remaining streams in the pipeline are destroyed, there is nowhere to send the data to, and the connection will never be released back into the pool, eventually locking up the application

this is bad and should not happen for pool.query(), I agree. At the same time if we have non pool connection or `.query()` called on a connection borrowed from pool maybe we should report error and allow next commands to be processed.

Same reasoning behind how it should behave if we start supporting query timeout - do we just ignore the data if it's coming after timeout? Do we keep connection open? To me healthy connection should not be killed automatically, timeout / downstream pipe error is "outside" problem to this connection and the user should decide what to do with connection.

**Comment by sidorares** [comment]
Would you be able to post very simple but full example illustrating connection not being released back to the pool @mdierolf ?

**Comment by mdierolf** [comment]
Yes, just need to get some free time to do so, I was going to write a unit test to trigger the behavior, and possibly a better fix.

Perhaps adding an abandoned connection feature to the pool would be a solution:
```
pool.on('abandonedConnection', connection => connection.destroy());
```

**Comment by sidorares** [comment]
I can't think of any scenarios why abandoned connection should not be returned back to the pool ( if it's still healthy )

**Comment by mdierolf** [comment]
The problem is, you don't know how long it will take to finish - in my case we have queries that have hundreds of megabytes of data returned. You might also have LOCK TABLE commands which are still holding table locks.

Here are a few options:
1) Do a TCP/IP teardown of the socket - typically this is 3 packets and a small amount of CPU time
2) Read the rest of the stream of data - this could be tens of thousands of packets
3) Let the user decide by exposing an interface such as above
4) Do nothing - maybe just document it in the docs

**Comment by erfanium** [comment]
Any news on this? I really needed to find a solution to this. Also I'm open to contribute and craft a PR

**Comment by mdierolf** [comment]
The solution is to update your app to retrieve the connection from the pool, and forcibly destroy the connection whenever you see fit.

For timeouts, it is best to set a statement timeout on your commands, and let mysql abort the query for you, and to try and avoid aborting them from the user side as much as possible.

You can use an abortcontroller or event emitter to call the connection.destroy() function when a user aborts or an app timeout occurs.

There isn't a single correct solution for the behavior the mysql client should take when a connection aborts from the user side, it is an application specific decision, so it would be a bad idea to make the client opinionated.

If anything, a section in the docs showing a few different variations of how to handle aborts from yhe server and user side would take a lot of the confusion out of this. Our business decided to wrap the node-mysql driver with an outer wrapper to handle aborts in a way that was consistent with our application needs

**Comment by erfanium** [comment]
@mdierolf 
When I use the pool in a fully managed way (i.e., not using `pool.getConnection()` and only using `pool.query(...)`), I don't expect a connection leak under any circumstances. It is the pool's responsibility to detect bad/dead connections and destroy them, not the application's.

**Comment by mdierolf** [comment]
The pool.query() interface is never going to be able to give you that strong guarantee, no matter how reasonable that sounds. 

That's why this issue still lurks around 5 years later.

**Comment by erfanium** [comment]
@mdierolf

```ts
const FIVE_MINUTES_MS = 5 * 60 * 1000;

type PoolConnection = mysql.PoolConnection & {
  _internalId: number;
  lastActiveTime: number;
};

function destroyPoolDeadConnections() {
  const now = Date.now();
  const pivot = now - FIVE_MINUTES_MS;

  const pool = client.pool as any;

  const all = pool._allConnections.toArray() as PoolConnection[];
  const frees = pool._freeConnections.toArray() as PoolConnection[];

  const deads = all.filter((c) => {
    const free = frees.includes(c);
    return !free && c.lastActiveTime < pivot;
  });

  for (const conn of deads) {
    conn.destroy();
    logger.info({
      msg: `[destroyPoolDeadConnections] id:${conn._internalId} threadId:${conn.threadId} destroyed`,
    });
  }
}

setInterval(destroyPoolDeadConnections, 10_000);

```
This is the code i use to destroy bad connections in my app. it's currently used in our critical systems with no issues so far. I think it's a good idea to embed this logic into the node-mysql pool. (disabled by default so no breaking changes)

I know this logic is pretty simple, but it should be possible to make it general-purpose enough to be helpful with most of use cases

**Comment by erfanium** [comment]
@mdierolf 

I read the thread again and I still don't understand what the problem is with propagating `stream.destroy()` to its underlying tcp connection (your first idea). How would it break something?
